src/eachobs.jl

```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
 end
 ```
 """
-function eachobs(data; buffer = false, batchsize::Int = -1, partial::Bool =true)
+function eachobs(
+        data;
+        buffer = false,
+        parallel = false,
+        shuffle = false,
+        batchsize::Int = -1,
+        partial::Bool = true,
+        executor = _default_executor())
     if batchsize > 0
         data = BatchView(data; batchsize, partial)
     end
-    if buffer === false 
-        gen = (getobs(data, i) for i in 1:numobs(data))
+
+    iter = if parallel
+        eachobsparallel(data; buffer, executor)
     else
-        if buffer === true && numobs(data) > 0
-            buffer = getobs(data, 1)
+        if buffer === false
+            EachObs(data)
+        elseif buffer === true
+            EachObsBuffer(data)
+        else
+            EachObsBuffer(data, buffer)
         end
-        gen = (getobs!(buffer, data, i) for i in 1:numobs(data))
     end
-    return gen
+
+    if shuffle
+        iter = ReshuffleIter(iter)
```

Comment:
I would prefer if we allow `EachObs` to contain a field `initialize` (or something) that gets called on the source data before iterating. Similar to how `DataLoader` calls `shuffleobs` but the function can be user specified.

This could allow resampling strategies every pass over the data (e.g. something that comes up in my colleagues research).

Reply:
this can be discussed in a future issue/pr

Reply:
Yeah, that's okay by me.